contracts/exchange-libs/contracts/src/LibFillResults.sol

```diff
@@ -45,17 +48,358 @@ contract LibFillResults is
+            leftOrder.takerAssetAmount,
+            leftTakerAssetAmountRemaining
+        );
+        uint256 rightTakerAssetAmountRemaining = rightOrder.takerAssetAmount.safeSub(rightOrderTakerAssetFilledAmount);
+        uint256 rightMakerAssetAmountRemaining = LibMath.safeGetPartialAmountFloor(
+            rightOrder.makerAssetAmount,
+            rightOrder.takerAssetAmount,
+            rightTakerAssetAmountRemaining
+        );
+
+        // Maximally fill the orders and pay out profits to the matcher in one or both of the maker assets.
+        if (shouldMaximallyFillOrders) {
+            matchedFillResults = _calculateMatchedFillResultsWithMaximalFill(
+                leftOrder,
+                rightOrder,
+                leftMakerAssetAmountRemaining,
+                leftTakerAssetAmountRemaining,
+                rightMakerAssetAmountRemaining,
+                rightTakerAssetAmountRemaining
+            );
+        } else {
+            matchedFillResults = _calculateMatchedFillResults(
+                leftOrder,
+                rightOrder,
+                leftMakerAssetAmountRemaining,
+                leftTakerAssetAmountRemaining,
+                rightMakerAssetAmountRemaining,
+                rightTakerAssetAmountRemaining
+            );
+        }
+
+        // Compute fees for left order
+        matchedFillResults.left.makerFeePaid = LibMath.safeGetPartialAmountFloor(
+            matchedFillResults.left.makerAssetFilledAmount,
+            leftOrder.makerAssetAmount,
+            leftOrder.makerFee
+        );
+        matchedFillResults.left.takerFeePaid = LibMath.safeGetPartialAmountFloor(
+            matchedFillResults.left.takerAssetFilledAmount,
+            leftOrder.takerAssetAmount,
+            leftOrder.takerFee
+        );
+
+        // Compute fees for right order
+        matchedFillResults.right.makerFeePaid = LibMath.safeGetPartialAmountFloor(
+            matchedFillResults.right.makerAssetFilledAmount,
+            rightOrder.makerAssetAmount,
+            rightOrder.makerFee
+        );
+        matchedFillResults.right.takerFeePaid = LibMath.safeGetPartialAmountFloor(
+            matchedFillResults.right.takerAssetFilledAmount,
+            rightOrder.takerAssetAmount,
+            rightOrder.takerFee
+        );
+
+        // Return fill results
+        return matchedFillResults;
+    }
+
     /// @dev Adds properties of both FillResults instances.
-    ///      Modifies the first FillResults instance specified.
-    /// @param totalFillResults Fill results instance that will be added onto.
-    /// @param singleFillResults Fill results instance that will be added to totalFillResults.
-    function _addFillResults(FillResults memory totalFillResults, FillResults memory singleFillResults)
+    /// @param fillResults1 The first FillResults.
+    /// @param fillResults2 The second FillResults.
+    /// @return The sum of both fill results.
+    function addFillResults(
```

Comment:
I'm a fan of this not being a destructive function. I think that we should make this the norm (if it isn't already).